File: app/src/main/java/com/karanchuk/roman/testtranslate/common/model/Translation.java
package com.karanchuk.roman.testtranslate.common.model;

import com.google.gson.annotations.SerializedName;

import java.util.List;

/**
 * Created by roman on 11.4.17.
 */

public class Translation {
    private String mNumber;

    @SerializedName("text")
    private String mText;

    @SerializedName("gen")
    private String mGen;

    @SerializedName("syn")
    private List<Synonym> mSynonyms;

    @SerializedName("mean")
    private List<Meaning> mMeanings;

    @SerializedName("ex")
    private List<Expression> mExpressions;

    private String mRepresentSynonyms;
    private String mRepresentMeanings;
    private String mRepresentExpressions;

    public Translation(final String number,
                       final List<Synonym> synonyms,
                       final List<Meaning> meanings,
                       final List<Expression> expressions,
                       final String representSynonyms) {
        mNumber = number;
        mSynonyms = synonyms;
        mMeanings = meanings;
        mExpressions = expressions;
        mRepresentSynonyms = representSynonyms;
    }

    public String getRepresentExpressions(){
        mRepresentExpressions = "";
        if (mExpressions != null) {
            for (Expression expression : mExpressions) {
                mRepresentExpressions = mRepresentExpressions.concat(expression.toString());
            }
        }
        int length = mRepresentExpressions.length();
        if (length >= 1){
            mRepresentExpressions = mRepresentExpressions.substring(0, length-1);
        }
        return mRepresentExpressions;
    }

    public String getRepresentMeanings() {
        if (mMeanings != null) {
            mRepresentMeanings = "(";
            for (Meaning meaning : mMeanings) {
                mRepresentMeanings = mRepresentMeanings.concat(meaning.toString() + ", ");
            }
        }
        return mRepresentMeanings.substring(0, mRepresentMeanings.length()-2) + ")";
    }

    public String getGen() {
        return mGen;
    }

    public void setGen(String gen) {
        mGen = gen;
    }

    public String getText() {
        return mText;
    }

    public void setText(String text) {
        mText = text;
    }

    public void setRepresentExpressions(String representExpressions) {
        mRepresentExpressions = representExpressions;
    }

    public void setRepresentMeanings(String representMeanings) {
        mRepresentMeanings = representMeanings;
    }

    public void setRepresentSynonyms(final String representSynonyms) {
        mRepresentSynonyms = representSynonyms;
    }

    public String getRepresentSynonyms(){
        if (mText != null){
            mRepresentSynonyms = mText;
        }
        if (mGen != null) {
            mRepresentSynonyms += " " + mGen.concat(", ");
        } else {
            if (mSynonyms != null){
                mRepresentSynonyms += ", ";
            }
        }
        if (mSynonyms != null) {
            for (Synonym synonym : mSynonyms) {
                if (synonym.getGen() != null) {
                    mRepresentSynonyms += synonym.toString() + " " + synonym.getGen() + ", ";
                } else {
                    mRepresentSynonyms += synonym.toString() + ", ";
                }
            }
        }
        int length = mRepresentSynonyms.length();
        if (mRepresentSynonyms.endsWith(", ")){
            mRepresentSynonyms = mRepresentSynonyms.substring(0, length - 2);
        } else if (mRepresentSynonyms.endsWith(",")){
            mRepresentSynonyms = mRepresentSynonyms.substring(0, length - 1);
        }
        return mRepresentSynonyms;
    }

    public List<Synonym> getSynonyms() {
        return mSynonyms;
    }

    public void setSynonyms(final List<Synonym> synonyms) {
        mSynonyms = synonyms;
    }

    public String getNumber() {
        return mNumber;
    }

    public void setNumber(final String number) {
        mNumber = number;
    }

    @Override
    public String toString() {
        String result = "";
        if (mNumber != null) {
            result = mNumber.concat(" ");
        }
        if (mSynonyms != null) {
            for (Synonym synonym : mSynonyms) {
                if (synonym.getGen() != null && !synonym.getGen().isEmpty()) {
                    result += synonym.toString() + " " + synonym.getGen() + ", ";
                } else {
                    result += synonym.toString() + ", ";
                }
            }
        }
        if (result.length() >= 2){
            result = result.substring(0, result.length()-2);
        }
        result = result.concat("\n").concat("(");
        if (mMeanings != null) {
            for (Meaning meaning : mMeanings) {
                result = result.concat(meaning.toString() + " ");
            }
        }
        if (result.length() >= 1){
            result = result.substring(0, result.length()-1);
        }
        result = result.concat(")").concat("\n");
        if (mExpressions != null) {
            for (Expression expression : mExpressions) {
                result = result.concat(expression.toString());
            }
        }
        return result;
    }

    public List<Meaning> getMeanings() {
        return mMeanings;
    }

    public void setMeanings(List<Meaning> meanings) {
        mMeanings = meanings;
    }

    public List<Expression> getExpressions() {
        return mExpressions;
    }

    public void setExpressions(List<Expression> expressions) {
        mExpressions = expressions;
    }
}

File: app/src/main/java/com/romankaranchuk/translator/data/database/model/Translation.kt
package com.romankaranchuk.translator.data.database.model

import com.romankaranchuk.translator.data.database.model.Synonym
import com.romankaranchuk.translator.data.database.model.Meaning
import com.google.gson.annotations.SerializedName

/**
 * Created by roman on 11.4.17.
 */
class Translation(
    var number: String?,
    @SerializedName("syn") var synonyms: List<Synonym>?,
    @SerializedName("mean") var meanings: List<Meaning>?,
    @SerializedName("ex") var expressions: List<Expression>?,
    private var mRepresentSynonyms: String
) {

    @SerializedName("text")
    var text: String? = null

    @SerializedName("gen")
    var gen: String? = null

    private var mRepresentMeanings: String? = null
    private var mRepresentExpressions: String? = null
    var representExpressions: String?
        get() {
            mRepresentExpressions = ""
            if (expressions != null) {
                for (expression in expressions!!) {
                    mRepresentExpressions = mRepresentExpressions + expression.toString()
                }
            }
            val length = mRepresentExpressions!!.length
            if (length >= 1) {
                mRepresentExpressions = mRepresentExpressions!!.substring(0, length - 1)
            }
            return mRepresentExpressions
        }
        set(representExpressions) {
            mRepresentExpressions = representExpressions
        }
    var representMeanings: String?
        get() {
            if (meanings != null) {
                mRepresentMeanings = "("
                for (meaning in meanings!!) {
                    mRepresentMeanings = "$mRepresentMeanings$meaning, "
                }
            }
            return mRepresentMeanings!!.substring(0, mRepresentMeanings!!.length - 2) + ")"
        }
        set(representMeanings) {
            mRepresentMeanings = representMeanings
        }
    var representSynonyms: String
        get() {
            if (text != null) {
                mRepresentSynonyms = text!!
            }
            if (gen != null) {
                mRepresentSynonyms += " " + gen + ", "
            } else {
                if (synonyms != null) {
                    mRepresentSynonyms += ", "
                }
            }
            if (synonyms != null) {
                for (synonym in synonyms!!) {
                    mRepresentSynonyms += if (synonym.gen != null) {
                        synonym.toString() + " " + synonym.gen + ", "
                    } else {
                        "$synonym, "
                    }
                }
            }
            val length = mRepresentSynonyms.length
            if (mRepresentSynonyms.endsWith(", ")) {
                mRepresentSynonyms = mRepresentSynonyms.substring(0, length - 2)
            } else if (mRepresentSynonyms.endsWith(",")) {
                mRepresentSynonyms = mRepresentSynonyms.substring(0, length - 1)
            }
            return mRepresentSynonyms
        }
        set(representSynonyms) {
            mRepresentSynonyms = representSynonyms
        }

    override fun toString(): String {
        var result = ""
        if (number != null) {
            result = number + " "
        }
        if (synonyms != null) {
            for (synonym in synonyms!!) {
                result += if (synonym.gen != null && !synonym.gen.isEmpty()) {
                    synonym.toString() + " " + synonym.gen + ", "
                } else {
                    "$synonym, "
                }
            }
        }
        if (result.length >= 2) {
            result = result.substring(0, result.length - 2)
        }
        result = "$result\n("
        if (meanings != null) {
            for (meaning in meanings!!) {
                result = "$result$meaning "
            }
        }
        if (result.length >= 1) {
            result = result.substring(0, result.length - 1)
        }
        result = "$result)\n"
        if (expressions != null) {
            for (expression in expressions!!) {
                result = result + expression.toString()
            }
        }
        return result
    }
}